### PR TITLE
Convert ajax to fetch

### PR DESCRIPTION
**Pull request by TheTrio**
## What this PR does

Addresses #4684 

The coverage report indicates that there are currently 12 requests which haven't been tested. 

1. [`fetchWikidataLabelsPromise`](https://github.com/WikiEducationFoundation/WikiEduDashboard/pull/4818/files#diff-7db6e393e6f74dbb08ce544acb55415ad34735fc73d58a5da18b9ba53fee4886)
2. [`article_graphs.jsx`](https://github.com/WikiEducationFoundation/WikiEduDashboard/pull/4818/files#diff-4fd0bc428b65148d07c6dbae3f39b909dcb827e52951dd7e4a2685cb0cd0f3c0)
3. [`updateSalesforceCredentialsPromise`](https://github.com/WikiEducationFoundation/WikiEduDashboard/pull/4818/files#diff-7ced3f635bf09bb5c531ecd01903b936ca5143a641f9dcfed64f8ab520940354)
4. [`grantAdminPromise`](https://github.com/WikiEducationFoundation/WikiEduDashboard/pull/4818/files#diff-7ced3f635bf09bb5c531ecd01903b936ca5143a641f9dcfed64f8ab520940354)
5. `createCustomFeedback`, `destroyCustomFeedback`, `greetStudents`, `dismissNotification`, `createBadWorkAlert`, `requestNewAccount`, `enableAccountRequests`, and `updateSalesforceRecord` all in [`api.js`](https://github.com/WikiEducationFoundation/WikiEduDashboard/pull/4818/commits/36d0df0f52bae1750ff88e5414879ea57e7cd6d7#diff-fe52ddf568c7cec6e7f56bc61c464f5b923852485bb723f0bb3fc3328c06e34e)

Apart from these, I haven't been able to test whether the `SentryLogger` works as expected. Its used in 2 places - `saveTimeline` and `saveCourse`, both of them in `api.js`

### CODE REVIEW

**Comment by TheTrio** [comment]
Not sure why that failed. I haven't changed anything related to syllabus upload. In fact, when I run the test locally it works fine 
image

**Comment by TheTrio** [comment]
@ragesoss I think that covers everything for this PR

**Comment by ragesoss** [comment]
Were you able to manually test all the ones that aren't covered in the rspec suite?

I'm not worried about the SentryLogger part. The actual network calls related to that are all handled by the Sentry SDK (via calls to Sentry.captureMessage) and the SentryLogger object is just used to hold data that we want to log. So as long as the requests that use Sentry are themselves working, I expect the Sentry part to be fine. (Breaking those would also be very low impact.)

**Comment by TheTrio** [comment]
I've tested `updateSalesforceCredentialsPromise` and `updateSalesforceCredentialsPromise` and they seem to work fine. As for others, like the feedback one, they are visible only if certain conditions are satisfied, so testing them will require a bit more effort. Haven't gotten around to doing that yet.

File: app/assets/javascripts/actions/wikidata_actions.js [hunk 1]
```diff
@@ -2,28 +2,27 @@ import { chunk, map, join, filter } from 'lodash-es';
 import * as types from '../constants';
 import logErrorMessage from '../utils/log_error_message';
 import CourseUtils from '../utils/course_utils';
+import request from '../utils/request';
+
 
 const wikidataApiBase = 'https://www.wikidata.org/w/api.php?action=wbgetentities&format=json&origin=*';
 
-const fetchWikidataLabelsPromise = (qNumbers) => {
+const fetchWikidataLabelsPromise = async (qNumbers) => {
```
Comment (ragesoss) [inline app/assets/javascripts/actions/wikidata_actions.js:10]: This appears to be broken. The `body` params are not being included.

I wrote a test that uses this (and breaks with your change): https://github.com/WikiEducationFoundation/WikiEduDashboard/commit/b9866a6071a842e10574c706364545ae9c064ba7

Comment (TheTrio) [reply app/assets/javascripts/actions/wikidata_actions.js:10]: Should be fixed now

**Comment by TheTrio** [comment]
I think I have found another problem. 

Currently, the `request` function adds a `X-CSRF-Token` only if the path is relative. However, it defines `relative` to mean not satisfying the regex `/http(s?)/g`. This works fine, until you have a request which has the word `http` or `https` in it. 

For example, the delete assignment request has a query parameter `sandboxUrl` which is set to `https%3A%2F%2Fwww.wikidata.org%2Fwiki%2FUser%3ADumbPotato21%2FQ16060473` or whatever the URL of the user is. 

image

This results in an unauthorized error 

image

To fix this, I've redefined it to be a relative path only if the URL doesn't _**start**_ with `http` or `https`. Hopefully this doesn't break anything else although I can't see how it would.

**Comment by ragesoss** [comment]
nice catch. how did you discover that problem?

**Comment by TheTrio** [comment]
Had no idea how to reproduce that wiki data bug so was just looking around, pressing buttons randomly. Found that I couldn't remove an item once I assigned it to a user. It was working on master so I knew it was a change I made that borked it. After that, it was just comparing the requests made by this branch and master.

**Comment by TheTrio** [comment]
I don't think this is related. It runs fine locally. 

image

**Comment by ragesoss** [comment]
@TheTrio Now that this is live, I've noticed a problem I missed before... when there is a server error, it returns HTML even for a json request. This used to result in an error notification like "Internal Server Error", but now instead it shows a JSON parsing error (because there's another error in the JS now when it tries to parse the HTML response as JSON).

For example, try adding `raise StandardError` to the first line of `CoursesController#update`, then editing the Description of a course and saving it.

Want to take a shot at fixing that? Any of the routes could potentially return an HTML error response.

**Comment by TheTrio** [comment]
If I have understood you correctly, all I would need to do is check if the response from the server is valid json. If it isn't, just show a generic error message(ie, internal server error)

Is that right?

**Comment by TheTrio** [comment]
Yeah that makes a lot more sense. I'll try this tomorrow.

**Pull request merged**